catalog/core/catalog-core-subscriptionstore/subscriptionstore-impl/src/main/java/org/codice/ddf/catalog/subscriptionstore/SubscriptionContainerImpl.java

```diff
@@ -0,0 +1,396 @@
+        return metadata;
+    }
+
+    /**
+     * {@inheritDoc}
+     * <p>
+     * Updates are performed by a delete first, then an insertion using the same id. The operations
+     * are identical to {@link #insert(Subscription, MarshalledSubscription, SubscriptionType)}
+     * and to {@link #delete(SubscriptionIdentifier)} with the exception of preserving the id in the
+     * {@link SubscriptionMetadata}.
+     */
+    @Override
+    public synchronized void update(Subscription subscription,
+            MarshalledSubscription marshalledSubscription, SubscriptionIdentifier identifier) {
+        validateSubscription(subscription);
+        validateSerialized(marshalledSubscription);
+        validateContainment(identifier, "update");
+
+        String subscriptionId = identifier.getId();
+        doDelete(subscriptionId);
+
+        SubscriptionMetadata metadata = new SubscriptionMetadata(identifier.getTypeName(),
+                marshalledSubscription.getFilter(),
+                marshalledSubscription.getCallbackAddress(),
+                subscriptionId);
+
+        doInsert(subscription, metadata);
+    }
+
+    /**
+     * {@inheritDoc}
+     * <p>
+     * Remove the subscription from the cache first, which will propogate the delete to the backing
+     * store. If an issue occurs during the delete, the subscription will <b>not</b> be unregistered.
+     */
+    @Override
+    public synchronized Subscription delete(SubscriptionIdentifier identifier) {
+        validateContainment(identifier, "delete");
+        String subscriptionId = identifier.getId();
+        CachedSubscription cachedSubscription = doDelete(subscriptionId);
+        return cachedSubscription.getSubscription()
+                .orElseThrow(() -> new SubscriptionStoreException("Could not get subscription. "));
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public synchronized boolean contains(SubscriptionIdentifier identifier) {
+        return get(identifier) != null;
+    }
+
+    /**
+     * Called when the OSGi framework registers a factory, typically from the container's client. It
+     * could be an endpoint, or any other subscription provider.
+     *
+     * @param factory the {@link SubscriptionFactory} that was registered and is now available.
+     */
+    public synchronized void bindFactory(SubscriptionFactory factory) {
+        if (factory == null) {
+            LOGGER.debug("Subscription container binding was given a null factory. ");
+            return;
+        }
+
+        if (factories.containsKey(factory.getTypeName())) {
+            throw new SubscriptionStoreException(
+                    "Duplicate factory registered. This is an API client error");
+        }
+
+        factories.put(factory.getTypeName(), factory);
+        LOGGER.debug("SubscriptionFactory registered for {}", factory.getTypeName());
+
+        StreamSupport.stream(subscriptionsCache.spliterator(), false)
```

Comment:
Factories appearing and disappearing will be limited to install and possibly rare configuration cases. I'm not sure if they are worth optimizing, but if we find a straightforward path to do so, let's take it.